File: 0189-rotate-array/0189-rotate-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    let n=nums.length;
  // console.log(n);
  k=k%n;

  reverseArr(nums,0,n-1);
  reverseArr(nums,0,k-1);
  reverseArr(nums,k,n-1);

  // console.log(nums);
};


var reverseArr= function(nums,start,end) {

  while(start<end){

    let temp=nums[end];
    nums[end]=nums[start];
    nums[start]=temp;

    start++;
    end--;
  }
}

File: 0189-rotate-array/0189-rotate-array.java
/* Explanation video :
https://www.youtube.com/watch?v=TYT5TJSfGlo
*/

class Solution {
    public void rotate(int[] nums, int k) {
        
        
        int n=nums.length;
        k=k%n;
        reverse(nums,0,n-1);
        reverse(nums,0,k-1);
        reverse(nums,k,n-1);
        
        
    }
    
     public void reverse(int[] nums,int start,int end){

        while(start<end){

            int temp=nums[end];
            nums[end]=nums[start];
            nums[start]=temp;

            start++;
            end--;
        }
     }
}